solvers/SimpleSOL/Diagnostics/momentum_conservation.hpp

```diff
@@ -0,0 +1,307 @@
+
+
+    return global_particle_momentum_transferred_1;
+  }   
+  
+  
+  
+   inline double compute_total_added_momentum_0() {
+    const double particle_momentum_added_0 =
+        this->particle_sys->total_particle_momentum_added[0];
+    double global_particle_momentum_added_0;
+    MPICHK(MPI_Allreduce(&particle_momentum_added_0, &global_particle_momentum_added_0, 1,
+                         MPI_DOUBLE, MPI_SUM,
+                         sycl_target->comm_pair.comm_parent));
+
+
+    return global_particle_momentum_added_0;
+  } 
+  
+  
+  
+  
+
+
+  inline double compute_fluid_0_momentum() {
+    return this->momentum_0_ions->Integral(this->momentum_0_ions->GetPhys()) ;
+  }
+  
+  inline double compute_fluid_1_momentum() {
+    return this->momentum_1_ions->Integral(this->momentum_1_ions->GetPhys()) ;
+  }
+
+  inline void compute_initial_fluid_0_momentum() {
+    if (!this->initial_fluid_0_momentum_computed) {
+      this->initial_0_momentum_fluid = this->compute_fluid_0_momentum();
+      this->initial_fluid_0_momentum_computed = true;
+    }
+  }
+  
+  inline void compute_initial_fluid_1_momentum() {
+    if (!this->initial_fluid_1_momentum_computed) {
+      this->initial_1_momentum_fluid = this->compute_fluid_1_momentum();
+      this->initial_fluid_1_momentum_computed = true;
+    }
+  }
+  
+  
+  
+  
+    inline void compute_initial_particle_0_momentum() {
+    if (!this->initial_particle_0_momentum_computed) {
+      this->initial_0_momentum_particles = this->compute_particle_momentum_0();
+      this->initial_particle_0_momentum_computed = true;
+    }
+  }
+  
+  inline void compute_initial_particle_1_momentum() {
+    if (!this->initial_particle_1_momentum_computed) {
+      this->initial_1_momentum_particles = this->compute_particle_momentum_1();
+      this->initial_particle_1_momentum_computed = true;
+    }
+  }
+  
+  
+  
+
+  inline double get_initial_fluid_0_momentum() {
+    NESOASSERT(this->initial_fluid_0_momentum_computed == true,
+               "initial x momentum not computed");
+    return this->initial_0_momentum_fluid;
+  }
+  
+  inline double get_initial_1_momentum() {
+    NESOASSERT(this->initial_fluid_1_momentum_computed == true,
+               "initial y momentum not computed");
+    return this->initial_1_momentum_fluid;
+  }
+  
+  
+  inline double get_initial_particle_0_momentum() {
+    NESOASSERT(this->initial_particle_0_momentum_computed == true,
+               "initial x momentum not computed");
+    return this->initial_0_momentum_particles;
+  }
+  
+  inline double get_initial_particle_1_momentum() {
+    NESOASSERT(this->initial_fluid_1_momentum_computed == true,
+               "initial y momentum not computed");
+    return this->initial_1_momentum_particles;
+  }
+
+  inline void compute(int step) {
+    if (momentum_recording_step > 0) {
+      if (step % momentum_recording_step == 0) {
+        const double momentum_particles_0 = this->compute_particle_momentum_0();
```

Comment:
Guessing these should be calls to `get_initial_*` and the calls to `get_initial_*` should in turn call `compute_initial_*` and/or call the compute methods in some setup/constructor.